### DISCUSSION TITLE
The "calling C from V" section is too brief

### DISCUSSION THREAD

**Discussion by emdete** [vlang/v#14637] [Ideas]
I want to access the g* family of libraries (glib, gmime, ...) and wonder how to use defines & macros provided by the headers. I didnt get even enums to work (their constants doesnt seem to have the type of the enum). 

Examples:

- The object model of glib has a safe cast macro like `GMIME_OBJECT()` which i wasnt able to use
- The library has defines for `TRUE` and `FALSE` which i wasnt able to use (same applies to `NULL` i assume where the sqlite sample uses 0)
- gmime has an enum `GMimeNewLineFormat` with `GMIME_NEWLINE_FORMAT_DOS` but the type of that is not `GMimeNewLineFormat` but `int` in V
- V generates `struct X` even if X is a typedef and i wasnt able to change that behaviour
- i did not get option/result types for error handling working, it v likes my code bu generates code that is not liked by the c compiler - are option types on c-structs possible at all?

Can the documentation of the "Calling C from V" section be enhanced to explain solutions?

**Comment by ylluminate** [comment]
@emdete have you had luck moving forward personally? If so, have you thought about contributing PRs to improve the section as and where you believe it could be improved?

**Comment by emdete** [comment]
@ylluminate Yes, meanwhile i learned alot. i put some doc&code together and would like to discuss it - what would be the best place?

**Reply by medvednikov** [reply]
just create a PR, we'll discuss it there

or on discord.gg/vlang

**Comment by emdete** [comment]
i dont use discord (why isnt the room bridged to matrix anymore?) 

point is: i have some questions, so for a PR its to early. 

maybe i just put it here. while playing around to understand this `C.` i came to this funny behaviour:

```
	x := C.strcpy
	eprintln("strcpy is $x")
	y := C.printf
	eprintln("printf is $y")
```

the first `eprintln` just gives a number and i thought ok, thats a c function pointer (is that true?). but the second gives "printf is fn (char, []voidptr) int". thats strange.. where does that come from? there even seems to be a `*` missing for the first parameter.

**Reply by div72** [reply]
About matrix: The bridges have been cut mostly due to spam. We added a [moderation bot](https://github.com/matrix-org/mjolnir) that shares banlists with other projects(like Python Community's Matrix, or the Matrix Foundation Moderated channels). So the spam should be much lower because of that, and if not I can just make the bot more ruthless by enabling some plugins. Irregardless, I am not sure if @medvednikov would be open to re-bridging.

About the question you asked: V assumes every non-defined `C.<name>` to be an int macro, so the first one's address of strcpy, while the second one's considered a C function so it also has a function type.

**Comment by emdete** [comment]
but shouldnt printf be a `fn (&char, []voidptr) int`?

**Reply by div72** [reply]
@emdete The definition is here looks correct to me: https://github.com/vlang/v/blob/8fa1e30dd2ab47a6b7dce1c987a58cca331f7ee1/vlib/builtin/cfns.c.v#L50

Perhaps the string representation is broken?

**Comment by emdete** [comment]
that's what i wrote till now:

https://gist.github.com/emdete/1b77c309062101446f03fa07ec29e9c6

it executable code to be sure it works (the header is missing). it's what i needed to do my integration, not sure if it's correct at all ;)
